### PR TITLE
Adding JSON extractor support

### PR DESCRIPTION
**Pull request by KaulSe**
I am not a professional GoLang developer. Might need adjustments (e.g. I used the package https://github.com/itchyny/gojq for `jq` like support).

Example:

```yaml
id: gitlab-public-repos

info:
  name: GitLab public repositories
  author: ldionmarcil
  severity: info
  tags: gitlab
requests:
  - method: GET
    path:
      - "{{BaseURL}}/api/v4/projects"
    headers:
      Cookie: _gitlab_session=

    extractors:
      - type: json
        part: body
        json:
          - '.[] | .id'
        name: projects
        group: 1
```

### CODE REVIEW

File: v2/pkg/protocols/http/operators_test.go [hunk 1]
```diff
@@ -168,6 +168,20 @@ func TestHTTPOperatorExtract(t *testing.T) {
...
+
+	t.Run("json", func(t *testing.T) {
+		extractor := &extractors.Extractor{
+			Type: "json",
+			Json: []string{".batters | .batter | .[] | .id"},
+		}
+		err = extractor.CompileExtractors()
+		require.Nil(t, err, "could not compile json extractor")
+
+		event["body"] = exampleJSONResponseBody
```
Comment (KaulSe) [inline v2/pkg/protocols/http/operators_test.go:180]: also not quite sure about this approach

**Comment by KaulSe** [comment]
also something to thing about; I am using `itemString := types.ToString(v)` and it returns something like this if the jq query is aiming at an object or similar;

```json
{
...
"extracted_results": [
    "map[avatar_url:<nil> full_path:gitlab-qa-sandbox-group/qa-test-2021-08-01-12-09-21-8c9ac8ebb3522246 id:1.2889589e+07 kind:group name:qa-test-2021-08-01-12-09-21-8c9ac8ebb3522246 parent_id:4.909902e+06 path:qa-test-2021-08-01-12-09-21-8c9ac8ebb3522246 web_url:https://gitlab.com/groups/gitlab-qa-sandbox-group/qa-test-2021-08-01-12-09-21-8c9ac8ebb3522246]",
...
]
...
}
```
with  
```yaml
id: gitlab-public-repos

info:
  name: GitLab public repositories
  author: ldionmarcil
  severity: info
  tags: gitlab
requests:
  - method: GET
    path:
      - "{{BaseURL}}/api/v4/projects"
    headers:
      Cookie: _gitlab_session=

    extractors:
      - type: json
        part: body
        json:
          - '.[] | .namespace' # <-- -- -- -- 
        name: projects
        group: 1
```

If we would transform `v` to a JSON string, it could result in problems in internal processing.
Given the JSON body:
```json
{
        "name": "test",
}
```
and the GoLang code: 
```golang
			bytes, err := json.Marshal(v)
			if err != nil {
				break
			}
			results[string(bytes)] = struct{}{}
```
and the JQ payload: `.name`.
results into `"test"` but if you want to use it in further requests, you need the unquoted version imo.

what do you think? the best option would be to dump objects as JSON strings with `json.Marshal(v)` and otherwise just use `types.ToString(v)` or do you have a better approach?

**Comment by ehsandeep** [comment]
Thank you for creating this PR @KaulSe. 

We were planning to enhance matcher/extractor support by adding this and a few additional kinds here - #584, and most likely will be adding this support for matchers as well. We will examine this PR and give feedback or update changes as necessary.

File: v2/pkg/protocols/http/operators_test.go (hunk 1 above)
Comment (Ice3man543) [reply v2/pkg/protocols/http/operators_test.go:180]: the test looks good to me, it performs the intended task

**Comment by KaulSe** [comment]
I added more tests and added a solution for this problem https://github.com/projectdiscovery/nuclei/pull/898#issuecomment-890510932

File: v2/pkg/operators/extractors/extract.go [hunk 2]
```diff
@@ -42,3 +44,34 @@ func (e *Extractor) ExtractKval(data map[string]interface{}) map[string]struct{}
 	}
 	return results
 }
+
+// ExtractJSON extracts key value pairs from a data map
+func (e *Extractor) ExtractJSON(corpus string) map[string]struct{} {
```
[the remark this answers is not in the document]
Comment (KaulSe) [reply v2/pkg/operators/extractors/extract.go:49]: I agree. Json.Unmarshal should solve that already

File: v2/pkg/operators/extractors/extract.go [hunk 3]
```diff
@@ -42,3 +44,34 @@ func (e *Extractor) ExtractKval(data map[string]interface{}) map[string]struct{}
...
+		iter := k.Run(jsonObj)
+		for {
+			v, ok := iter.Next()
+			if !ok {
+				break
+			}
+			if _, ok := v.(error); ok {
+				break
+			}
+			itemString := types.ToString(v)
```
[the remark this answers is not in the document]
Comment (KaulSe) [reply v2/pkg/operators/extractors/extract.go]: actually, I found another solution for that. Inspiration came from the GitHub client;
https://github.com/cli/cli/blob/09b09810dd812e3ede54b59ad9d6912b946ac6c5/pkg/export/template.go#L72
https://github.com/cli/cli/blob/09b09810dd812e3ede54b59ad9d6912b946ac6c5/pkg/export/filter.go#L38

**Review by Ice3man543 — APPROVED** [review]
Looks good to me! Working as intended.

**Comment by Ice3man543** [comment]
This has been merged in the dev version, soon we'll update documentation as well to reflect these new additions. Thank you again for the PR @KaulSe!

**Pull request merged**